File: src/org/dyndns/fzoli/rccar/bridge/socket/ControllerSideMessageProcess.java
package org.dyndns.fzoli.rccar.bridge.socket;

import java.io.Serializable;
import org.dyndns.fzoli.rccar.ConnectionKeys;
import org.dyndns.fzoli.rccar.model.PartialBaseData;
import org.dyndns.fzoli.rccar.model.bridge.ControllerStorage;
import org.dyndns.fzoli.rccar.model.bridge.HostStorage;
import org.dyndns.fzoli.rccar.model.bridge.StorageList;
import org.dyndns.fzoli.rccar.model.controller.ControllerData;
import org.dyndns.fzoli.socket.handler.SecureHandler;

/**
 * A vezérlő-kliens szerver oldali üzenetfeldolgozója.
 * @author zoli
 */
public class ControllerSideMessageProcess extends BridgeMessageProcess implements ConnectionKeys {

    /**
     * A vezérlő kliens munkamenete a szerveren.
     */
    private ControllerStorage storage;

    public ControllerSideMessageProcess(SecureHandler handler) {
        super(handler);
    }

    /**
     * Kapcsolódáskor létrehozza/frissíti a vezérlő munkamenetét és ha nincs leválasztott járműje, elküldi a járműlistát a vezérlőnek,
     * egyébként visszaállítja a járművet és elküldi a járműadatokat.
     * A vezérlő oldalon ezáltal megjelenik a járműválasztó ablak vagy a régebben kiválasztott járműhöz tartozó felület.
     */
    @Override
    protected void onStart() {
        storage = StorageList.createControllerStorage(this);
        if (!storage.hasDisconnectedHost()) {
            sendMessage(StorageList.createHostList(getRemoteCommonName()));
        }
        else {
            storage.restoreDisconnectedHost();
            sendMessage(storage.createControllerData());
        }
    }

    /**
     * Ha részadat érkezik a klienstől, átadja a munkamenet üzenetfogadójának.
     */
    @Override
    protected void onMessage(Serializable o) {
        if (o instanceof PartialBaseData) {
            PartialBaseData<ControllerData, ?> pd = (PartialBaseData) o;
            if (storage != null) storage.getReceiver().update(pd);
        }
    }

}

File: cpp/RCServer/ControllerSideMessageProcess.cpp
/* 
 * File:   TestMessageProcess.cpp
 * Author: zoli
 * 
 * Created on 2013. szeptember 14., 22:41
 */

#include "ControllerSideMessageProcess.h"
#include "ControllerData.h"
#include "StorageList.h"

#include <iostream>

ControllerSideMessageProcess::ControllerSideMessageProcess(SSLHandler* handler) : MessageProcess(handler) {
    storage = NULL;
}

void ControllerSideMessageProcess::onStart() {
    storage = (ControllerStorage*) StorageList::createControllerStorage(this);
    if (!storage->hasDisconnectedHost()) {
        HostList msg = StorageList::createHostList(getSocket()->getClientName());
        sendMessage(&msg);
    }
    else {
        storage->restoreDisconnectedHost();
        ControllerData msg = storage->createControllerData();
        sendMessage(&msg);
    }
}

void ControllerSideMessageProcess::onMessage(Message* msg) {        
    if (storage) storage->getReceiver()->update(msg);
    delete msg;
}

void ControllerSideMessageProcess::onUnknownMessage(UnknownMessage* msg) {
    if (Command::isCommand(msg)) {
        Command cmd(msg);
        if (storage) storage->onCommand(&cmd);
    }
    else {
        std::cout << "Unknown controller message: " << msg->getClassName() << " - " << msg->getDefinition() << std::endl;
    }
    delete msg;
}
